### PR TITLE
feat: add .ngsummary.ts files to support AOT unit tests

### PR DESCRIPTION
**Pull request by tbosch**
Design doc: https://docs.google.com/document/d/1VmTkz0EbEVSWfEEWEvQ5sXyQXSCvtMOw4t7pKU-jOwc/edit?usp=sharing

### CODE REVIEW

File: packages/compiler/src/jit/compiler.ts [hunk 1]
```diff
@@ -106,14 +127,21 @@ export class JitCompiler implements Compiler {
...
-    // of imported modules.
-    // However, for runtime compilation, we want to transitively compile all modules,
-    // so we also need to call loadNgModuleDirectiveAndPipeMetadata for all nested modules.
-    ngModule.transitiveModule.modules.forEach((localModuleMeta) => {
-      loadingPromises.push(this._metadataResolver.loadNgModuleDirectiveAndPipeMetadata(
-          localModuleMeta.reference, isSync));
+    const mainNgModule = this._metadataResolver.getNgModuleMetadata(mainModule) !;
+    // Note: for runtime compilation, we want to transitively compile all modules,
+    // so we also need to load the declared directives / pipes for all nested modules.
+    const transJitModules = this._filterJitIdentifiers(mainNgModule.transitiveModule.modules);
```
Comment (chuckjaz) [inline packages/compiler/src/jit/compiler.ts]: Consider `transitiveJitModules`

File: packages/compiler/src/summary_resolver.ts [hunk 2]
```diff
@@ -14,11 +15,25 @@ export interface Summary<T> {
...
+  abstract getLibraryFileName(fileName: string): string|null;
+  abstract resolveSummary(reference: T): Summary<T>|null;
+  abstract getSymbolsOf(filePath: string): T[];
+  abstract getImportAs(reference: T): T;
+  abstract addSummary(summary: Summary<T>): void;
+}
+
 @CompilerInjectable()
-export class SummaryResolver<T> {
+export class NoopSummaryResolver implements SummaryResolver<Type<any>> {
```
Comment (chuckjaz) [inline packages/compiler/src/summary_resolver.ts]: Consider `JitSummaryResolver` since this is no longer a noop.

**Review by chuckjaz — APPROVED** [review]
Approved with some nits to fix.

File: packages/compiler/src/jit/compiler.ts (hunk 1 above)
Comment (tbosch) [inline packages/compiler/src/jit/compiler.ts]: Done

File: packages/compiler/src/jit/compiler.ts [hunk 3]
```diff
@@ -106,14 +127,21 @@ export class JitCompiler implements Compiler {
...
+    transJitModules.forEach((nestedNgModule) => {
+      const moduleMeta = this._metadataResolver.getNgModuleMetadata(nestedNgModule) !;
```
Comment (tbosch) [inline packages/compiler/src/jit/compiler.ts]: Done

File: packages/compiler/src/jit/compiler.ts [hunk 4]
```diff
@@ -364,3 +392,18 @@ class ModuleBoundCompiler implements Compiler {
    */
   clearCacheFor(type: Type<any>) { this._delegate.clearCacheFor(type); }
 }
+
+
+function flattenSummaries(fn: () => any[], out: CompileTypeSummary[] = []): CompileTypeSummary[] {
+  if (!fn) {
```
Comment (tbosch) [inline packages/compiler/src/jit/compiler.ts]: Done

File: packages/compiler/src/aot/static_symbol_resolver.ts [hunk 5]
```diff
@@ -103,6 +111,18 @@ export class StaticSymbolResolver {
           this.getStaticSymbol(baseImportAs.filePath, baseImportAs.name, staticSymbol.members) :
           null;
     }
+    const summarizedFileName = stripSummaryForJitFileSuffix(staticSymbol.filePath);
+    if (summarizedFileName !== staticSymbol.filePath) {
+      const summarizedName = stripSummaryForJitName(staticSymbol.name);
```
Comment (tbosch) [inline packages/compiler/src/aot/static_symbol_resolver.ts]: Done

File: packages/compiler/src/aot/static_symbol_resolver.ts [hunk 6]
```diff
@@ -128,7 +148,7 @@ export class StaticSymbolResolver {
     // cause the metadata for an factory file to be loaded which doesn't exist.
```
Comment (tbosch) [inline packages/compiler/src/aot/static_symbol_resolver.ts]: Done

File: packages/compiler/src/aot/static_symbol_resolver.ts [hunk 7]
```diff
@@ -138,6 +158,20 @@ export class StaticSymbolResolver {
...
   }
 
+  /**
+   * Converts a file path to a module name that can be used as an `import`.
+   */
+  fileNameToModuleName(importedFilePath: string, containingFilePath: string): string|null {
+    if (importedFilePath === containingFilePath) {
+      return null;
+    }
+    let result = this.knownFileNameToModuleNames.get(importedFilePath) || null;
```
Comment (tbosch) [inline packages/compiler/src/aot/static_symbol_resolver.ts]: Done

File: packages/compiler/src/aot/static_symbol_resolver.ts [hunk 8]
```diff
@@ -236,12 +270,10 @@ export class StaticSymbolResolver {
         const name = unescapeIdentifier(metadataKey);
 
         const symbol = this.getStaticSymbol(filePath, name);
-        let importSymbol: StaticSymbol|undefined = undefined;
         if (metadata['importAs']) {
```
Comment (tbosch) [inline packages/compiler/src/aot/static_symbol_resolver.ts]: Done

File: packages/compiler/src/summary_resolver.ts (hunk 2 above)
Comment (tbosch) [inline packages/compiler/src/summary_resolver.ts]: Done

**Review by IgorMinar — APPROVED** [review]
lgtm for api changes

**Comment by matsko** [comment]
Landed as https://github.com/angular/angular/commit/547c3634733d8fb49ff88730cc3d6f13a2ac9fb8

**Pull request closed without merging**